File: src/main/java/de/gesellix/docker/client/filesocket/HostnameEncoder.java
package de.gesellix.docker.client.filesocket;

import okio.ByteString;

import java.util.ArrayList;
import java.util.List;

class HostnameEncoder {

    /**
     * @see java.net.IDN
     */
    private final static Integer MAX_LABEL_LENGTH = 63;
    private final static Integer MAX_HOSTNAME_LENGTH = MAX_LABEL_LENGTH * 4;

    String encode(String toEncode) {
        String encoded = ByteString.encodeUtf8(toEncode).hex();
        if (encoded.length() > MAX_LABEL_LENGTH && encoded.length() < MAX_HOSTNAME_LENGTH) {
            List<String> labels = new ArrayList<>();
            int labelCount = (int) Math.ceil(encoded.length() / MAX_LABEL_LENGTH.doubleValue());
            for (int step = 0; step < labelCount; step++) {
                int from = step * MAX_LABEL_LENGTH;
                int to = from + MAX_LABEL_LENGTH;
                labels.add(encoded.substring(from, Math.min(to, encoded.length())));
            }
            return String.join(".", labels);
        }
        return encoded;
    }

    String decode(String toDecode) {
        String decoded = toDecode;
        if (toDecode.contains(".")) {
            String[] labels = toDecode.split("\\.");
            decoded = String.join("", labels);
//            decoded = elements.dropLastWhile { it.isEmpty() }.toTypedArray().joinToString("");
        }
        return ByteString.decodeHex(decoded).utf8();
    }
}

File: src/main/kotlin/de/gesellix/docker/client/filesocket/HostnameEncoder.kt
package de.gesellix.docker.client.filesocket

import okio.ByteString

class HostnameEncoder {

    companion object {

        /**
         * @see java.net.IDN
         */
        const val MAX_LABEL_LENGTH = 63
        const val MAX_HOSTNAME_LENGTH = MAX_LABEL_LENGTH * 4
    }

    fun encode(toEncode: String): String {
        val encoded = ByteString.encodeUtf8(toEncode).hex()
        if (encoded.length > MAX_LABEL_LENGTH && encoded.length < MAX_HOSTNAME_LENGTH) {
            val labelCount = Math.ceil((encoded.length / MAX_LABEL_LENGTH).toDouble()).toInt()
            val labels = (0..labelCount).fold(ArrayList<String>()) { substrings, step ->
                val from = step * MAX_LABEL_LENGTH
                val to = from + MAX_LABEL_LENGTH
                substrings.add(encoded.substring(from, Math.min(to, encoded.length)))
                substrings
            }
            return labels.joinToString(".")
        }
        return encoded
    }

    fun decode(toDecode: String): String {
        var decoded = toDecode
        if (toDecode.contains(".")) {
            decoded = toDecode.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray().joinToString("")
        }
        return ByteString.decodeHex(decoded).utf8()
    }
}
